### PR TITLE
Phase 2: Depth × Width Scaling Grid (8 parallel experiments)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The single biggest Phase 1 constraint was **n_layers=1**, chosen purely for throughput (more epochs in 30 minutes). With 3 hours of training, deeper and wider models should significantly outperform the current architecture. The original Transolver used 5 layers; our 1-layer model was a speed-accuracy tradeoff. More layers = more representational capacity for complex flow interactions, especially tandem wake dynamics.

## Instructions

**CRITICAL: All experiments must change these constants at the top of train.py:**
```python
MAX_TIMEOUT = 180.0  # 3 hours instead of 30 min
MAX_EPOCHS = 500     # Allow more epochs
```

Run 8 experiments in parallel, one per GPU. Each experiment varies `n_layers` and/or `n_hidden` in `model_config`, with schedules scaled proportionally to the estimated epoch count.

### Common schedule adjustments (scale proportionally to est. epochs)
For each configuration, estimate epochs = `180 * 60 / epoch_time_seconds`. Then set:
- `warmup total_iters = max(10, int(0.1 * est_epochs))`
- `cosine T_max = est_epochs - warmup`
- `ema_start_epoch = int(0.6 * est_epochs)`
- Temperature annealing epoch: change `if epoch >= 50` to `if epoch >= int(0.7 * est_epochs)`
- Progressive volume ramp: change `if epoch < 40` to `if epoch < int(0.4 * est_epochs)` and scale denominator
- Tandem curriculum: change `if epoch < 10` to `if epoch < int(0.1 * est_epochs)`
- Noise annealing: change `epoch / 60` to `epoch / int(0.6 * est_epochs)` everywhere

### GPU 0: n_layers=2, n_hidden=192 (baseline deeper)
`CUDA_VISIBLE_DEVICES=0 python train.py --wandb_name "frieren/p2-L2-H192" --wandb_group "phase2-depth-width" --agent frieren`

### GPU 1: n_layers=3, n_hidden=192
`CUDA_VISIBLE_DEVICES=1 python train.py --wandb_name "frieren/p2-L3-H192" --wandb_group "phase2-depth-width" --agent frieren`

### GPU 2: n_layers=4, n_hidden=192
`CUDA_VISIBLE_DEVICES=2 python train.py --wandb_name "frieren/p2-L4-H192" --wandb_group "phase2-depth-width" --agent frieren`

### GPU 3: n_layers=2, n_hidden=256
`CUDA_VISIBLE_DEVICES=3 python train.py --wandb_name "frieren/p2-L2-H256" --wandb_group "phase2-depth-width" --agent frieren`

### GPU 4: n_layers=2, n_hidden=320
`CUDA_VISIBLE_DEVICES=4 python train.py --wandb_name "frieren/p2-L2-H320" --wandb_group "phase2-depth-width" --agent frieren`

### GPU 5: n_layers=4, n_hidden=256
`CUDA_VISIBLE_DEVICES=5 python train.py --wandb_name "frieren/p2-L4-H256" --wandb_group "phase2-depth-width" --agent frieren`

### GPU 6: n_layers=2, n_hidden=192, slice_num=64
`CUDA_VISIBLE_DEVICES=6 python train.py --wandb_name "frieren/p2-L2-S64" --wandb_group "phase2-depth-width" --agent frieren`

### GPU 7: n_layers=2, n_hidden=192, slice_num=96
`CUDA_VISIBLE_DEVICES=7 python train.py --wandb_name "frieren/p2-L2-S96" --wandb_group "phase2-depth-width" --agent frieren`

**Important**: The `fun_dim` in model_config stays the same (X_DIM - 2 + 2 + 32 = 56). Only `n_hidden`, `n_layers`, `n_head`, and `slice_num` change.

## Baseline
Current best metrics (Phase 1, noam HEAD, n_layers=1, n_hidden=192):
| Metric | Value |
|--------|-------|
| val/loss | 0.833 |
| p_in (in-dist surf_p MAE) | 17.9 Pa |
| p_oodc (OOD cond surf_p MAE) | 14.0 Pa |
| p_tan (tandem surf_p MAE) | 36.7 Pa |
| p_re (OOD Re surf_p MAE) | 27.5 Pa |

---

### Implementation note
All 8 schedule/architecture hyperparameters were made CLI-configurable in `train.py` (new flags: `--n_layers`, `--n_hidden`, `--n_head`, `--slice_num`, `--warmup_iters`, `--cosine_t_max`, `--ema_start_epoch`, `--temp_anneal_epoch`, `--vol_ramp_end`, `--tandem_curr_end`, `--noise_anneal_denom`, `--hard_mining_epoch`). This allows all 8 experiments to run in parallel with proportionally scaled schedules without separate train.py copies. End-of-run visualization errored (missing Fourier PE in vis code — pre-existing bug); training metrics were fully logged to W&B.

### Summary table

| Config | val/loss | p_in (Pa) | p_oodc (Pa) | p_tan (Pa) | p_re (Pa) | Epochs | Peak GB | W&B run |
|--------|----------|-----------|-------------|------------|-----------|--------|---------|---------|
| **Baseline** (L1-H192-S48) | 0.833 | 17.9 | 14.0 | 36.7 | 27.5 | — | — | — |
| L2-H192-S48 | **0.729** | 14.7 | 11.0 | 36.9 | 25.6 | 246 | 26.5 | re9l7qfk |
| L3-H192-S48 | 0.735 | 15.5 | **10.7** | 36.6 | 25.8 | 164 | 33.1 | 9bu9et39 |
| L4-H192-S48 | 0.942 | 19.8 | 14.2 | 44.2 | 27.6 | 142 | 39.6 | c9lucb3x |
| L2-H256-S48 | 0.733 | **14.6** | 10.9 | 37.2 | 25.7 | 195 | 34.2 | 8z7lipfj |
| L2-H320-S48 | 0.733 | 14.9 | 10.9 | 36.7 | 25.9 | 154 | 42.4 | 93hbequs |
| L4-H256-S48 | 1.003 | 19.7 | 16.4 | 45.6 | 29.2 | 96 | 51.1 | 62x4gznj |
| **L2-H192-S64** | **0.729** | 15.0 | 11.0 | 36.5 | 25.7 | 224 | **26.2** | zaq7s3h3 |
| L2-H192-S96 | 0.732 | 15.1 | 11.2 | **36.2** | 26.0 | 198 | 29.6 | fziz60ks |

### In-distribution MAE detail (best checkpoint, val_in_dist)
| Config | surf_Ux | surf_Uy | surf_p | vol_Ux | vol_Uy | vol_p |
|--------|---------|---------|--------|--------|--------|-------|
| L2-H192 | 1.74 m/s | 0.67 m/s | 14.7 Pa | 0.70 m/s | 0.23 m/s | 14.7 Pa |
| L2-S64  | 1.75 m/s | 0.67 m/s | 15.0 Pa | 0.70 m/s | 0.24 m/s | 14.7 Pa |

### What happened

**n_layers=2 is the clear winner.** Every 2-layer config beat the 1-layer baseline by a large margin (~12–14% on val/loss: 0.729–0.733 vs 0.833). The hypothesis was correct — depth was the bottleneck. Going from 1→2 layers gave the biggest single gain of Phase 1+2.

**n_layers=4 failed badly.** Both L4-H192 (0.942) and L4-H256 (1.003) are significantly *worse* than baseline. At 4 layers, the models are slower per epoch (~72–92 s vs ~42–57 s for 2-layer), so they only reached 96–142 epochs vs 154–246 for 2-layer. The deeper architectures need more time to converge than 3 hours allows. The train/loss for L4-H192 (0.579) vs L2-H192 (0.203) confirms the 4-layer models are still far from convergence — this is a training budget issue, not a fundamental architecture failure.

**Width scaling shows diminishing returns.** L2-H256 (0.733) and L2-H320 (0.733) barely improve over L2-H192 (0.729) while using 34–42 GB vs 26.5 GB. Not worth the memory premium.

**Slice scaling is mildly helpful at S64.** L2-S64 (val/loss=0.7294) edges out L2-H192 (0.7295) by a hair and uses slightly *less* memory (26.2 GB vs 26.5 GB), making it the best efficiency/performance tradeoff. S96 (0.732) slightly underperforms S64 — too many slices may over-fragment the routing.

**Tandem pressure is stuck.** All 2-layer models hover at 36.2–37.2 Pa for p_tan, barely different from the 36.7 Pa baseline. The tandem ceiling is the known SURFACE_IDS=(5,6) limitation (missing foil 2 boundary nodes), not model capacity. Depth and width don't help here.

**Pressure improvements elsewhere are solid.** p_in improved 18% (17.9→14.6 Pa best), p_oodc improved 24% (14.0→10.7 Pa best with L3-H192), p_re improved 7% (27.5→25.6 Pa).

### Suggested follow-ups

1. **L4 with 6h budget**: L4-H192 and L4-H256 are clearly undertrained (train/loss still 0.58 at end). Run them for 6 hours — if they converge past the 2-layer configs, deeper architectures may be worth it for a final model.
2. **L2-H256-S64 combined**: Try n_layers=2, n_hidden=256, slice_num=64 — combining modest width boost with routing improvement. The individual gains are small but may combine.
3. **L3-H192 deserves more time**: At 164 epochs it beats all 2-layer configs on p_oodc (10.7 Pa) and is essentially tied on val/loss. 6 hours may pull it ahead.
4. **Tandem surface fix**: Adding boundary ID 7 (foil 2 surface) to SURFACE_IDS would directly improve p_tan — this is the primary ceiling.
5. **Recommended next default**: **L2-S64** (run zaq7s3h3) — best val/loss, lowest memory, good convergence. Recommended base for Phase 3.

### CODE REVIEW

**Comment by tcapelle** [comment]
## Round 1 Results — Depth × Width Scaling Grid

Solid results across the board, frieren! The depth scaling hypothesis is confirmed.

### Results Summary

| GPU | Config | val/loss | p_in | p_oodc | p_tan | p_re | vs Baseline |
|-----|--------|---------|------|--------|-------|------|-------------|
| 6 | L2-S64 | 0.729 | 15.0 | 11.0 | 36.5 | 25.7 | -12% |
| 0 | **L2-H192** | **0.729** | **14.7** | 11.0 | 36.9 | **25.6** | **-12%** |
| 7 | L2-S96 | 0.732 | 15.1 | 11.2 | 36.2 | 26.0 | -12% |
| 4 | L2-H320 | 0.733 | 14.9 | 10.9 | 36.7 | 25.9 | -12% |
| 3 | L2-H256 | 0.733 | 14.6 | 10.9 | 37.2 | 25.7 | -12% |
| 1 | L3-H192 | 0.735 | 15.5 | 10.7 | 36.6 | 25.8 | -12% |
| 2 | L4-H192 | 0.942 | 19.8 | 14.2 | 44.2 | 27.6 | Not converged |
| 5 | L4-H256 | 0.989 | 19.7 | 15.3 | 45.9 | 28.5 | Not converged |

### Key Findings
1. **n_layers=2 is the sweet spot** — all L2 variants perform similarly (0.729-0.733)
2. **Width barely matters**: H192, H256, H320 all converge to ~0.733. No need to go wider.
3. **n_layers=3 is competitive** (0.735) but no better than L2
4. **n_layers=4 didn't converge** (0.94-0.99) — too slow, not enough epochs in 3h
5. **slice_num=64 and 96** are similar to 48 — finer routing doesn't help much
6. **L2-H192 is the efficiency winner** — best results with smallest model, fastest training

### Verdict: MERGE L2-H192 as the baseline depth change
